### PR TITLE
Support TLS configuration of AerospikeCluster Client through env variables.

### PR DESCRIPTION
**Pull request by skolosov-snap**
Feeding Client Certs to the operator through files on operator's filesystem.
New logic is following:
CRD now has a new section AerospikeClientCertSourceSpec containing client certs information for connecting to Aerospike.
If AerospikeClientCertSource is set then operator reads certs/key from the files, otherwise, read them from the secret.
This should be reloadable since operator does not store/cache certs or AS client connections.

Following validation is added:
- if tls-authenticate-client is "false", then we do not require to provide client cert, however you still can provide CA certs.
- if tls-authenticate-client is "any", then client cert has to be provided, but you don't need to provide tlsClientName
- otherwise you have to provide both client cert and tlsClientName (which will be added to tls-authenticate-client if missed)

Also validation added to make sure tls-authenticate-client contains name (CN or DNS from SAN) from operator's cert if configured.

### CODE REVIEW

**Comment by arrowplum** [comment]
Discussions have been continuing in Slack channel.

File: api/v1alpha1/aerospikecluster_types.go [hunk 1]
```diff
@@ -70,10 +70,39 @@ type AerospikeClusterSpec struct {
...
+	// This name will be added to tls-authenticate-client list by the operator
+	TLSClientName string `json:"tlsClientName"`
+	CertSource    `json:",inline"`
+}
+
+// CertSource represents the source of Aerospike ClientCert for Operator.
+// Only one of its members may be specified.
+type CertSource struct {
+	// +optional
+	SecretName *string `json:"secretName,omitempty"`
```
Comment (sud82) [inline api/v1alpha1/aerospikecluster_types.go:90]: secretName and namespace can be put in a struct like secretSource. CertSource should be like 
```
CertSource {
  *source1
  *source2
  *source3
}
```
In the future, we may also like to add some more fields inside any source. For e.g. I suggested the below schema for secret, not sure if you missed it or deliberately decided to skip it

```
type AerospikeSecretCertSource struct {
    SecretName string `json:"secretName"`
	CaCertsFilename    string `json:"caCertFilename"`
	// +optional
	ClientCertFilename string `json:"clientCertFilename"`
	// +optional
	ClientKeyFilename  string `json:"clientKeyFilename"`
}
```
ClientCertFilename, ClientKeyFilename are optional if tls-authenticate-client is "any"
https://docs.aerospike.com/docs/reference/configuration/#tls-authenticate-client

**Review by skolosov-snap — COMMENTED** [review]
> Discussions have been continuing in Slack channel.

@arrowplum, I'm not in this discussion. The last update on a slack was week ago. Is there any other thread for this discussion I'm not aware of?

**Review by sud82 — CHANGES_REQUESTED** [review]

File: api/v1alpha1/aerospikecluster_validating_webhook.go [hunk 2]
```diff
@@ -208,6 +215,25 @@ func (r *AerospikeCluster) validate(aslog logr.Logger) error {
...
+	} else if !IsTLS(configSpec) && clientCertSpec != nil {
+		return fmt.Errorf("operatorClientCertSpec is configured but TLS is not")
+	}
+	if clientCertSpec == nil {
+		return nil
+	}
+	if (clientCertSpec.SecretName == nil) == (clientCertSpec.CertPathInOperator == nil) {
+		return fmt.Errorf("either SecretName (%s) or certPathInOperator (%v) must be set in operatorClientCertSpec but not both",
+			*clientCertSpec.SecretName, *clientCertSpec.CertPathInOperator)
+	}
```
[the remark this answers is not in the document]
Comment (skolosov-snap) [reply api/v1alpha1/aerospikecluster_validating_webhook.go:250]: Nice catch. Refactored.

**Review by skolosov-snap — COMMENTED** [review]
I believe when tls-authenticate-client is "any" then client cert still needs to be used. because it's still mTLS but Aerospike server just doesn't validate name provided in client cert.

New logic is following:
- if tls-authenticate-client is "false", then we do not require to provide client cert, however you still can provide CA certs.
- tls-authenticate-client is "any", then client cert has to be provided, but you don't need to provide tlsClientName
- otherwise you have to provide both client cert and tlsClientName (which will be added to tls-authenticate-client if missed)

**Review by sud82 — CHANGES_REQUESTED** [review]

File: api/v1alpha1/aerospikecluster_types.go [hunk 3]
```diff
@@ -70,10 +70,73 @@ type AerospikeClusterSpec struct {
...
+	// +optional
+	ClientCertFilename string `json:"clientCertFilename,omitempty"`
+	// +optional
+	ClientKeyFilename string `json:"clientKeyFilename,omitempty"`
+}
+
+// AerospikeCertPathInOperatorSource contain configuration for certificates used by operator to connect to aerospike cluster.
+// All paths are on operator's filesystem.
+type AerospikeCertPathInOperatorSource struct {
+	// +optional
```
[the remark this answers is not in the document]
Comment (skolosov-snap) [reply api/v1alpha1/aerospikecluster_types.go:111]: Yes, caCertsPath is optional. If it's omitted then system CA certs are used by default.

File: api/v1alpha1/aerospikecluster_types.go [hunk 4]
```diff
@@ -70,10 +70,73 @@ type AerospikeClusterSpec struct {
...
+	ClientCertPath string `json:"clientCertPath,omitempty"`
+	// +optional
+	ClientKeyPath string `json:"clientKeyPath,omitempty"`
+}
+
+func (c *AerospikeOperatorClientCertSpec) IsClientCertConfigured() bool {
+	return (c.SecretCertSource != nil && c.SecretCertSource.ClientCertFilename != "") ||
+		(c.CertPathInOperator != nil && c.CertPathInOperator.ClientCertPath != "")
+}
```
[the remark this answers is not in the document]
Comment (skolosov-snap) [reply api/v1alpha1/aerospikecluster_types.go:122]: SecretCertSource.ClientCertFilename and ertPathInOperator.ClientCertPath are optional. You can use Sources to configure CA certs only (i.e. when tls-authenticate-client is "false"). Moreover, we have a validation which checks that exactly one of these two are not nil (see next method, first check). So "(c.SecretCertSource != nil || c.CertPathInOperator != nil )" will always be true.

**Comment by skolosov-snap** [comment]
Fixed tests.

**Review by sud82 — APPROVED** [review]

**Pull request merged**